rustls/src/crypto/aws_lc_rs/hpke.rs

```diff
@@ -0,0 +1,846 @@
+use alloc::boxed::Box;
+use alloc::sync::Arc;
+use alloc::vec::Vec;
+use core::fmt::{self, Debug, Formatter};
+
+use aws_lc_rs::aead::{
+    self, Aad, BoundKey, Nonce, NonceSequence, OpeningKey, SealingKey, UnboundKey, NONCE_LEN,
+};
+use aws_lc_rs::agreement;
+use aws_lc_rs::cipher::{AES_128_KEY_LEN, AES_256_KEY_LEN};
+use aws_lc_rs::digest::{SHA256_OUTPUT_LEN, SHA384_OUTPUT_LEN, SHA512_OUTPUT_LEN};
+use zeroize::Zeroize;
+
+use crate::crypto::aws_lc_rs::tls13::{
+    RING_HKDF_HMAC_SHA256, RING_HKDF_HMAC_SHA384, RING_HKDF_HMAC_SHA512,
+};
+use crate::crypto::hpke::{
+    EncapsulatedSecret, Hpke, HpkeOpener, HpkePrivateKey, HpkePublicKey, HpkeSealer, HpkeSuite,
+    HpkeSuites,
+};
+use crate::crypto::tls13::{expand, Hkdf, HkdfExpander};
+use crate::msgs::enums::{HpkeAead, HpkeKdf, HpkeKem};
+use crate::msgs::handshake::HpkeSymmetricCipherSuite;
+use crate::{Error, OtherError};
+
+/// `HpkeAwsLcRs` holds the concrete instantiations of the algorithms specified by the [HpkeSuite].
+pub struct HpkeAwsLcRs<const KEY_SIZE: usize, const KDF_SIZE: usize> {
+    suite: HpkeSuite,
+    dh_kem: &'static DhKem<KDF_SIZE>,
+    hkdf: &'static dyn Hkdf,
+    aead: &'static aead::Algorithm,
+}
+
+impl<const KEY_SIZE: usize, const KDF_SIZE: usize> HpkeAwsLcRs<KEY_SIZE, KDF_SIZE> {
+    /// See RFC 9180 §5.1 "Creating the Encryption Context"[^0].
+    ///
+    /// [0]: https://www.rfc-editor.org/rfc/rfc9180.html#section-5.1
+    fn key_schedule(
+        &self,
+        shared_secret: KemSharedSecret<KDF_SIZE>,
+        info: &[u8],
+    ) -> Result<(AeadKey<KEY_SIZE>, NonceSchedule), Error> {
+        // Note: we use an empty IKM for the `psk_id_hash` and `secret` labelled extractions because
+        // there is no PSK ID in base mode HPKE.
+
+        let suite_id = self.hpke_suite_id();
+        let psk_id_hash = labeled_extract(self.hkdf, &suite_id, None, Label::PskIdHash, &[]);
+        let info_hash = labeled_extract(self.hkdf, &suite_id, None, Label::InfoHash, info);
+        let key_schedule_context = [
+            &[0][..], // base mode (0x00)
+            psk_id_hash.expand_block(&[]).as_ref(),
+            info_hash.expand_block(&[]).as_ref(),
+        ]
+        .concat();
+
+        let key = AeadKey(labeled_expand::<KEY_SIZE>(
+            &suite_id,
+            labeled_extract(
+                self.hkdf,
+                &suite_id,
+                Some(&shared_secret.0),
+                Label::Secret,
+                &[],
+            ),
+            Label::Key,
+            &key_schedule_context,
+        ));
+
+        // Note: All supported AEADs use the same NONCE_LEN.
+        let base_nonce = labeled_expand::<NONCE_LEN>(
+            &suite_id,
+            labeled_extract(
```

Comment:
Looks like this is the same as the one for `key`. Extract?

Reply:
Extracted.